src/SFML/Graphics/Shape.cpp

```diff
@@ -143,6 +142,57 @@ FloatRect Shape::getGlobalBounds() const
     return getTransform().transformRect(getLocalBounds());
 }
 
+Vector2f Shape::getGeometricCenter() const
+{
+    const auto count = getPointCount();
+
+    switch (count)
+    {
+        case 0:
+            assert(false);
+            return Vector2f{};
+        case 1:
+            return getPoint(0);
+        case 2:
+            return (getPoint(0) + getPoint(1)) / 2.f;
+        default: // more than two points
+            Vector2f centroid;
+            float    twiceArea = 0;
+
+            auto previousPoint = getPoint(count - 1);
+            for (std::size_t i = 0; i < count; ++i)
+            {
+                const auto  currentPoint = getPoint(i);
+                const float product      = previousPoint.cross(currentPoint);
+                twiceArea += product;
+                centroid += (currentPoint + previousPoint) * product;
+
+                previousPoint = currentPoint;
+            }
+
+            if (twiceArea != 0.f)
+            {
+                return centroid / 3.f / twiceArea;
+            }
+
+            // Fallback for no area - find the midpoint between the two points furthest apart
+            auto minPoint = getPoint(0);
+            auto maxPoint = minPoint;
+            for (std::size_t i = 1; i < count; ++i)
+            {
+                const auto currentPoint = getPoint(i);
+                if (currentPoint.x < minPoint.x)
+                    minPoint.x = currentPoint.x;
+                else if (currentPoint.x > maxPoint.x)
+                    maxPoint.x = currentPoint.x;
+                if (currentPoint.y < minPoint.y)
+                    minPoint.y = currentPoint.y;
+                else if (currentPoint.y > maxPoint.y)
+                    maxPoint.y = currentPoint.y;
```

Comment:
Can this be simplified with `std::min` and `std::max`?

Reply:
Or `std::clamp`...

Reply:
Not sure how clamp could come in handy here. The point between the min and max is not what we’re trying to change. We’re trying to change the min and max themselves.

Reply:
> Can this be simplified with `std::min` and `std::max`?

Do you mean `minPoint.x = std::min(minPoint.x, currPoint.x);` etc.? Sure. I don't like it, but if you want it that way, I'll change it.

> Or `std::clamp` ...

I don't see how I could clamp here.

Reply:
> > Do you mean `minPoint.x = std::min(minPoint.x, currPoint.x);` etc.?
> 
> Yes. It's more expressive than rewriting that same logic by hand.

It's not exactly the same logic though. In the current version, if the first condition is `true` it will skip the second. Minor detail perhaps, but that's why I don't like using `min`/`max` in these situations.

Would it ok to make it expressive by hiding it like this:

    template<class T>
    std::pair<T, T> setMinMax(const T valueToTest, T& currentMin, T& currentMax) {
        if (valueToTest > currentMax) currentMax = valueToTest;
        else if (valueToTest < currentMin) currentMin = valueToTest;
    }

The code in the function with then become

            for (std::size_t i = 1; i < count; ++i)
            {
                const auto currentPoint = getPoint(i);
                setMinMax(currentPoint.x, minPoint.x, maxPoint.x);
                setMinMax(currentPoint.y, minPoint.y, maxPoint.y);
            }

If not, I'll go with `min`/`max`.

Reply:
> It's not exactly the same logic though.

The result is the same, right?

> Would it ok to make it expressive by hiding it like this:

This abstraction is not as intuitive to me and feels excessively complicated when all we need are a few calls to `min` and `max` functions.

Reply:
Ok, `min`/`max` it is then :-)